### ISSUE TITLE
[no-unused-vars] Allow unused type parameters when merging interface with class that has type parameters

### ISSUE THREAD

**Issue by ChocolateLoverRaj** [typescript-eslint/typescript-eslint#4098]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

Repl.it: https://replit.com/@Programmerraj/typescript-eslint-bug#index.ts
```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

```TS
interface Pineapple<T> {
  run?(): void
}

abstract class Pineapple<T> {
  abstract value: T
}

export default Pineapple
```

**Expected Result**
No lint error

`1:21  error  'T' is defined but never used  @typescript-eslint/no-unused-vars`

**Use case**
https://stackoverflow.com/a/59488269/11145447

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.3.0` |
| `@typescript-eslint/parser`        | `5.3.0` |
| `TypeScript`                       | `4,4,4` |
| `ESLint`                           | `8.2.0` |
| `node`                             | `12.22.6` |

**Comment by JoshuaKGoldberg** [comment]
Fun bug!

Note that `@typescript-eslint/no-unused-vars` does not have `requiresTypeChecking` enabled, so as long as this can be fixed without incurring a breaking change to require type checking, that'd be great.

**Comment by bradzacher** [comment]
Declaration merging an interface into a class is *bad practice* because it causes TS to create **unchecked** properties/methods on that class.

In the provided example it's "safe" because the method is ***explicitly*** marked as optional. However there's no reason you can't just... not mark the method as optional, leading to a method TS **thinks** is defined on the class, but actually isn't - which will of course crash at runtime.

It's one that historically we've not supported because
a) this pattern isn't good because it's unsafe
b) it's really really uncommon because it's unsafe
c) it can't be handled across files without type information.
d) this effect is weirdly for an engineer to understand - it's unused but required so it would be marked as "used" due to "actions at a distance".

TBH I don't think it's a good idea for us to support this.

**Comment by ChocolateLoverRaj** [comment]
@bradzacher so my issue is a duplicate of the one you linked. Should I continue the conversation in this issue or the other issue?

**Comment by bradzacher** [comment]
we can discuss here, if there's anything you'd like to add.

**Comment by ChocolateLoverRaj** [comment]
> Declaration merging an interface into a class is _bad practice_

@bradzacher why can't not merging interfaces and classes by a seperate rule instead of having to disable `no-unused-vars`?

**Comment by bradzacher** [comment]
> why can't not merging interfaces and classes by a seperate rule instead of having to disable no-unused-vars?

That could definitely be a separate rule!

However the issue is (c) above:
> c) it can't be handled across files without type information.

We can handle declaration merging within the same file only, and cannot handle across files - which is really the most common way to use declaration merging now a days.

In the case of "declaration merging within the same file" it's additionally really, *really* hard to handle declaration merging to completeness.

For example - it's **somewhat** doable to handle declarations when they're inline at the same level, like your example. But did you know it's also valid to do [declaration merging across namespaces?](https://www.typescriptlang.org/play?#code/HYQwtgpgzgDiDGEAEAxA9mpBvAUE-SEAHjGgE4AuSAlsBRGQGYLIBCIZ2eBPMZaMAFxIoFMrQDmAbm74AvjgU5QkWC1QYuPYqUpJ4AGxBQoSdpywKl8NMFFJGmgLxJgEAO4a0AOnMAKAEoZRx8+ARkcIA). This case is much, much harder to handle.

Which means that we could provide a solution which works specifically for one single case, and fails for all other cases - but then you have an inconsistent experience for users - why does the rule handle one case but not another that (to the user) is identical?
And if the user doesn't understand declaration merging (which a lot of users do not, because it's an advanced usecase), then this sort of "magical" handling can look like a false negative - which can be confusing.

So for the combination of the above reasons: very low usage across TS codebases, difficult implementation, incomplete implementation, etc - we opt to not support declaration merging unless it really adds value the rule.

**Comment by ChocolateLoverRaj** [comment]
@bradzacher so do you mean that adding this feature would take too much effort vs the benefits of implementing it?

**Comment by bradzacher** [comment]
Correct. It would take a decent amount of effort to implement and (IMO) the negatives (like asymmetric handling of declaration merging cases) outweigh the positives (supporting a rare case that very, very few people run into).
